functionalization: fix bug with multiple views of same base

bdhirsh opened the pull request:
welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795



Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527
* #76358
* #77714
* #77713
* #77132
* #77716
* #77130
* __->__ #77129
* #77128
* #77125
* #77126

bdhirsh committed:
functionalization: fix bug with multiple views of same base

[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -194,10 +194,8 @@ void FunctionalTensorWrapper::sync_() {
-    regenerate_from_base();
-  }
+  apply_updates();
+  regenerate_from_base();
```
Now you get to explain, in the code here, why skipping regeneration when there are no updates is wrong ;)

ezyang reviewed (approved):
I figured it out looking at the test, haha. Nice catch.

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix bug with multiple views of same base"

welp, I realized my "perf improvement" from https://github.com/pytorch/pytorch/pull/75819 was wrong. This originally came up because `functionalize()` was sometimes emit an unnecessary `view_copy`, but there's a more correct fix that I can make inside of functorch, which I have here: https://github.com/pytorch/functorch/pull/795






[ghstack-poisoned]
